processor/maskprocessor/processor.go

```diff
@@ -0,0 +1,187 @@
+	return nil
+}
+
+// processLogs masks incoming logs.
+func (p *processor) processLogs(_ context.Context, ld plog.Logs) (plog.Logs, error) {
+	for i := 0; i < ld.ResourceLogs().Len(); i++ {
+		resource := ld.ResourceLogs().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeLogs().Len(); j++ {
+			scope := resource.ScopeLogs().At(j)
+			for k := 0; k < scope.LogRecords().Len(); k++ {
+				logs := scope.LogRecords().At(k)
+				logs.Attributes().Range(p.maskAttrsFunc)
+				p.maskValue(bodyField, logs.Body())
+			}
+		}
+	}
+
+	return ld, nil
+}
+
+// processTraces masks incoming traces.
+func (p *processor) processTraces(_ context.Context, td ptrace.Traces) (ptrace.Traces, error) {
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		resource := td.ResourceSpans().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeSpans().Len(); j++ {
+			scope := resource.ScopeSpans().At(j)
+			for k := 0; k < scope.Spans().Len(); k++ {
+				spans := scope.Spans().At(k)
+				spans.Attributes().Range(p.maskAttrsFunc)
+			}
+		}
+	}
+
+	return td, nil
+}
+
+// processMetrics masks incoming metrics.
+func (p *processor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	for i := 0; i < md.ResourceMetrics().Len(); i++ {
+		resource := md.ResourceMetrics().At(i)
+		resource.Resource().Attributes().Range(p.maskResourceFunc)
+		for j := 0; j < resource.ScopeMetrics().Len(); j++ {
+			scope := resource.ScopeMetrics().At(j)
+			for k := 0; k < scope.Metrics().Len(); k++ {
+				metrics := scope.Metrics().At(k)
+				switch metrics.Type() {
+				case pmetric.MetricTypeSum:
+					p.processSum(metrics.Sum())
+				case pmetric.MetricTypeGauge:
+					p.processGauge(metrics.Gauge())
+				case pmetric.MetricTypeSummary:
+					p.processSummary(metrics.Summary())
+				}
+			}
+		}
+	}
+
+	return md, nil
+}
+
+// processSum masks a sum metric.
+func (p *processor) processSum(sum pmetric.Sum) {
+	for i := 0; i < sum.DataPoints().Len(); i++ {
+		sum.DataPoints().At(i).Attributes().Range(p.maskAttrsFunc)
+	}
+}
+
+// processGauge masks a gauge metric.
+func (p *processor) processGauge(gauge pmetric.Gauge) {
+	for i := 0; i < gauge.DataPoints().Len(); i++ {
+		gauge.DataPoints().At(i).Attributes().Range(p.maskAttrsFunc)
+	}
+}
+
+// processSummary masks a summary metrics.
+func (p *processor) processSummary(summary pmetric.Summary) {
+	for i := 0; i < summary.DataPoints().Len(); i++ {
+		summary.DataPoints().At(i).Attributes().Range(p.maskAttrsFunc)
+	}
+}
+
+// maskValue masks a pcommon.Value.
+func (p *processor) maskValue(field string, value pcommon.Value) {
+	for _, excludeField := range p.cfg.Exclude {
+		if field == excludeField {
+			return
+		}
+	}
+
+	switch value.Type() {
+	case pcommon.ValueTypeMap:
+		maskFunc := p.createMaskFunc(field)
+		value.Map().Range(maskFunc)
+	case pcommon.ValueTypeStr:
+		p.maskString(value)
+	}
```

Comment:
Is it worth supporting the "bytes" type here? I don't know of any receiver that uses it, probably only relevant if a user is using OTLP right now.

Also, maybe we want to iterate slices as well?

Reply:
I stubbed out handling bytes and it's a bit nasty as you have to rewrite the buffer byte by byte. I'm also not 100% confident it's appropriate to convert the bytes to a string then regex match on it and assume everything worked as intended. A string is more much straight forward. 

I think because the majority use case here is likely logs we can omit bytes and add later if a customer asks. @jmwilliams89 would like your input on this please.

As far as slices go I believe we can at least search a slice for strings and apply masking to those.

Reply:
>  I'm also not 100% confident it's appropriate to convert the bytes to a string then regex match on it and assume everything worked as intended.

I was thinking that you would just use the regex functions that work on bytes (e.g. [Match](https://pkg.go.dev/regexp#Regexp.Match) and [ReplaceAll](https://pkg.go.dev/regexp#Regexp.ReplaceAll)), but if it's not actually as straight-forward, then it probably isn't a big deal to omit.

Reply:
Yes we can do that. I'm not sure if blindly applying a regex to a range of bytes is always safe in this case. Maybe it's being overly cautious but I'm not sure if we can always assume the bytes can be treated as a string.